### DISCUSSION TITLE
How to add node selector?

### DISCUSSION THREAD

**Discussion by LoganZhao1997** [ds4dm/ml4co-competition#52] [Q&A]
Hi!

I am wondering whether I can create an environment for both node selection and branching? Specifically, I want to use my own node selection policy and branching policy to get a primal solution. However, it looks like that the ecole's environment only includes Branching.

Best regards,
Logan Zhao

**Comment by gasse** [comment]
Hi @LoganZhao1997 ,

Unfortunately Ecole currently does not support node selection, and does not support either mixing two different control problems into a single environment. Both are features we have envisioned already, and would like to have in Ecole, but unfortunately they do not exist yet.

Still, you don't necessarily need Ecole to achieve what you want. I believe you should be able to mix node selection + variable selection with [PySCIPOpt](https://github.com/scipopt/PySCIPOpt), by implementing your own [branching rule](https://github.com/scipopt/PySCIPOpt/blob/master/src/pyscipopt/branchrule.pxi) and [node selector](https://github.com/scipopt/PySCIPOpt/blob/master/src/pyscipopt/nodesel.pxi).

See also the following ressources:
 - https://www.scipopt.org/doc/html/group__NODESELECTORS.php
 - https://www.scipopt.org/doc/html/NODESEL.php
 - https://www.scipopt.org/doc/html/group__BRANCHINGRULES.php
 - https://www.scipopt.org/doc/html/BRANCH.php
 
 Hope that helps,
 
 Best,
 Maxime

**Reply by LoganZhao1997** [reply]
Thank you for your answer!
